Adopt Destination to Trigger

capri-xiyue opened the pull request:
Adopt Destination to Trigger 
Since https://github.com/knative/eventing/issues/1918 needs to adopt Destination in multiple resources like Trigger, Source, Parallel, Sequence, Subscription. To avoid huge code review, I will create multiple PRs (one PR for one resource or related resources) to avoid huge code review task

## Proposed Changes

- Adopt Destination to Subscriber in Trigger, fixed unit tests, unit tests and e2e tests in local passed
-
-

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
```

lionelvillard commented on config/300-trigger.yaml:
```diff
@@ -77,27 +77,29 @@ spec:
-                  minLength: 1
-                ref:
+              anyOf:
+                - uri:
```
should be properties ?

capri-xiyue replied:
Is anyOf better since the logic is at least one here?

grantr replied:
the `anyOf` here has the same effect as a single object, since the anyOf's options don't overlap.

capri-xiyue commented:
One question:

capri-xiyue commented:
Temporarily pause the progress of this issue since we have not made an decision for the Destination API. We're going to make the Destination a priority in our API task force to make sure that we can provide a same user experience. Will resume progress on this issue when we make a decision for the Destination API

grantr commented on pkg/reconciler/trigger/trigger.go:
```diff
@@ -201,9 +203,14 @@ func (r *Reconciler) reconcile(ctx context.Context, t *v1alpha1.Trigger) error {
 		if err := trackAddressable(*t.Spec.Subscriber.Ref); err != nil {
```
We might not need this trackAddressable anymore, since uriResolver does that. cc @n3wscott

capri-xiyue replied:
trackInNamespace use ensureTracking() to ensure that there is an informer watching and sending events to tracker for the concrete GVK. It also ensures that there is the corresponding lister for that informer. In the func URIFromDestination of uriResolver, it doesn't use any similar function like ensureTracking(). I'm not sure whether we can remove the  trackAddressable here.

capri-xiyue replied:
Is there anyway that I can do some manual test to see whether it works after we delete the trackAddressable? Should I do some changes for the resource of subscriber?

n3wscott replied:
You can subscribe to an object and have it change the resolved uri

capri-xiyue replied:
@n3wscott  I don't quite understand your meaning of "You can subscribe to an object and have it change the resolved uri". Could you give me some example?

capri-xiyue replied:
Do you mean I can do the manual test by letting Trigger subscribe to an object and then change the URI of the object?

capri-xiyue replied:
@grantr  @n3wscott  I did some manual tests after I deleted trackAddressable. I created Trigger which needs a subscriber `event-display`.At first it failed because the subscriber which the trigger needs is not present. Then I created the service `event-display`. The Trigger is able to reconcile again and becomes ready.

However, if i delete `event-display`, the status of Trigger won't change. It is still ready and Subscriber URI of status still exists.  I'm not sure whether this is the result we want. If it is a bug, the bug is not introduced by this PR. I tested the master branch. It's same. I checked the code, it is because we never change Subscriber URI when there is any change to the subscriber.

grantr replied:
I think that is not the result we want, but if the bug exists in master branch, we should create an issue to track it.

n3wscott commented:
I think you should revert the lock file changes out of this PR.

capri-xiyue commented:
> I think you should revert the lock file changes out of this PR.
I think I made some mistake when I resolved the conflicts with master branch regarding " Gopkg.lock". If I reverted the whole lock file, the code doesn't compile because currently the Gopkg.lock of master branch doesn't pick the latest version of knative pkg. Now I reverted "Gopkg.lock" to the version of master branch and then run `dep ensure -update knative.dev/pkg`. I think it should work.

grantr commented:
FYI @capri-xiyue is on vacation so I'm taking over this PR.

Harwayne commented on pkg/reconciler/trigger/trigger.go:
```diff
@@ -196,14 +198,19 @@ func (r *Reconciler) reconcile(ctx context.Context, t *v1alpha1.Trigger) error {
-		if err := trackAddressable(*t.Spec.Subscriber.Ref); err != nil {
-			return fmt.Errorf("unable to track changes to Subscriber.Ref: %v", err)
+	if t.Spec.Subscriber == nil {
+		return fmt.Errorf("Subscriber cannot be nil")
```
```suggestion
		return errors.New("subscriber cannot be nil")
```

Harwayne commented on pkg/reconciler/trigger/trigger.go:
```diff
@@ -196,14 +198,19 @@ func (r *Reconciler) reconcile(ctx context.Context, t *v1alpha1.Trigger) error {
+	if t.Spec.Subscriber.Ref != nil {
+		// To call URIFromDestination(dest apisv1alpha1.Destination, parent interface{}), dest.Ref must have a Namespace
+		// If there is no Namespace defined in dest.Ref, we will use the Namespace of Trigger as the Namespace of dest.Ref
+		if t.Spec.Subscriber.Ref.Namespace == "" {
```
Add a comment about why we can ignore the deprecated fields.

n3wscott commented:
/lgtm

Harwayne commented on pkg/reconciler/trigger/trigger.go:
```diff
@@ -196,14 +198,19 @@ func (r *Reconciler) reconcile(ctx context.Context, t *v1alpha1.Trigger) error {
+	if t.Spec.Subscriber.Ref != nil {
+		// To call URIFromDestination(dest apisv1alpha1.Destination, parent interface{}), dest.Ref must have a Namespace
+		// If there is no Namespace defined in dest.Ref, we will use the Namespace of Trigger as the Namespace of dest.Ref
+		if t.Spec.Subscriber.Ref.Namespace == "" {
```
As far as I can tell, historically we did not allow Namespace to be set here. When would we expect this to not be the empty string?

grantr replied:
I'm not sure. It looks like the previous code always used the Trigger's namespace, so we should probably do that here.

grantr commented:
/approve

This needs a test or two of the new destination functionality, similar to the ones in https://github.com/knative/eventing/pull/2089, but those can be added in a future PR.
